src/funcx_common/messagepack/protocol_versions/proto1.py

```diff
@@ -0,0 +1,128 @@
+  - if a payload defines fields which are not included in the current message
+    definitions, they are silently ignored
+
+Therefore, new fields may be added to the message classes, but they must not be
+required until all messages contain those fields.
+"""  # noqa: E501
+from __future__ import annotations
+
+import dataclasses
+import json
+import typing as t
+
+from ..common import Message
+from ..exceptions import InvalidMessagePayloadError, UnrecognizedMessageTypeError
+from ..message_types import ALL_MESSAGE_CLASSES
+from ..protocol import MessagePackProtocol
+
+# protocol version and reserved byte as a byte array
+_VERSION_BYTE = (1).to_bytes(1, byteorder="big", signed=False)
+_RESERVED_BYTE = (0).to_bytes(1, byteorder="big", signed=False)
+_STANDARD_HEADER = _VERSION_BYTE + _RESERVED_BYTE
+
+
+def _field_required(x: dataclasses.Field[t.Any]) -> bool:
+    # the following check is misreported by mypy as a non-overlapping check
+    # see: https://github.com/python/mypy/issues/12124
+    return (
+        x.default is dataclasses.MISSING
+        and x.default_factory is dataclasses.MISSING  # type: ignore[comparison-overlap]
+    )
+
+
+def _check_required_fields(
+    data: dict[str, t.Any], message_class: type[Message]
+) -> None:
+    required_message_fields = {
+        x.name for x in dataclasses.fields(message_class) if _field_required(x)
+    }
+    data_keys = set(data.keys())
+    if not required_message_fields <= data_keys:
+        raise InvalidMessagePayloadError(
+            f"message body of type {message_class.message_type} was missing required "
+            "fields: "
+            f"data_keys={data_keys}, required_message_fields={required_message_fields}"
+        )
+
+
+def _filter_data_to_fields(
+    data: dict[str, t.Any], message_class: type[Message]
+) -> dict[str, t.Any]:
+    message_fields = {x.name for x in dataclasses.fields(message_class)}
+    return {k: v for k, v in data.items() if k in message_fields}
+
+
+class MessagePackProtocolV1(MessagePackProtocol):
+    MESSAGE_TYPE_MAP: dict[str, type[Message]] = {}
+
+    def pack(self, message: Message) -> bytes:
+        body = {
+            "message_type": message.message_type,
+            "data": message.get_json_safe_dict(),
+        }
+        # encode() converts to bytes, but no characters will actually be altered because
+        # ensure_ascii is being used
+        return (
+            _STANDARD_HEADER
+            + json.dumps(body, separators=(",", ":"), ensure_ascii=True).encode()
+        )
+
+    def unpack(self, buf: bytes) -> Message:
+        # strip the two byte header
+        body = buf[2:]
+        envelope = json.loads(body)
+        if "message_type" not in envelope:
+            raise InvalidMessagePayloadError(
+                "cannot unpack message with no message_type"
+            )
+        if "data" not in envelope:
+            raise InvalidMessagePayloadError("cannot unpack message with no data")
+
```

Comment:
implicitly ignoring any other keys in the message here, right?

Reply:
Yes, I should specify this as well.

It's starting to make me think... there are three things which could reasonably either be ignored or raise errors:
- unknown field in envelope (here)
- unknown message_class
- unknown field in message body

Perhaps I should start using three bits of the flags byte I defined?
```python
raise_on_unknown_envelope_field = flags & 1
ignore_unknown_message_class = flags & 2
raise_on_unknown_fields = flags & 4
```
?

That way, the defaults are all defined, but it will be possible to put a message on the wire in a few different modes.

I could also specify these behaviors as an optional field or fields in the envelope.